### PR TITLE
Datecmp

### PR DESCRIPTION
**Pull request by bserdar**
JS Date comparisons must use date.valueOf()

### CODE REVIEW

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java [hunk 1]
```diff
@@ -219,20 +229,25 @@ private Block translateFieldComparison(Context ctx,FieldComparisonExpression que
...
             if(rFieldMd instanceof ArrayField) {
                 arrayFieldName=rfieldName;
                 simpleFieldName=lfieldName;
                 op=query.getOp().invert();
+                isDate=((ArrayField)rFieldMd).getElement().getType() instanceof DateType && lFieldMd.getType() instanceof DateType;
             } else {
                 arrayFieldName=lfieldName;
                 simpleFieldName=rfieldName;
                 op=query.getOp();
+                isDate=((ArrayField)lFieldMd).getElement().getType() instanceof DateType && rFieldMd.getType() instanceof DateType;
```
Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java]: Didn't we just confirm that lFieldMd is NOT an ArrayField?

**Comment by dcrissman** [comment]
@bserdar - Just for reference purposes, could you speak to why we need this?

Also I don't understand the change to the unit test. Is there an existing test case for this?

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java (hunk 1 above)
Comment (bserdar) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java]: At this point, only one of lField or rField is an ArrayField. The if statement is for rField being an array, and this is in the else, so at this point, lField is an array.

**Comment by bserdar** [comment]
This particular PR is to deal with Date comparison in Javascript code. It
turns out if you compare date values using <, <=, >, >=, it compares date
values, but for == and !=, it compares date objects, so if the field is a
date, we compare field.valueOf() instead of just field.

The change in the unit test simply removes one of the temp fields that we
no longer generate.

File: mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java (hunk 1 above)
Comment (dcrissman) [inline mongo/src/main/java/com/redhat/lightblue/mongo/crud/js/JSQueryTranslator.java]: Gotcha. I missed the nuance of the 'l' and 'r' prefixes.

**Comment by bserdar** [comment]
'l' is 'left'
'r' is 'right'

**Pull request merged**